algorithms/kernel/dbscan/oneapi/dbscan_dense_default_batch_ucapi_impl.i

```diff
@@ -0,0 +1,511 @@
+    KernelRange global_range(queueBlockSize, _maxWorkgroupSize);
+
+    KernelNDRange range(2);
+    range.global(global_range, &st);
+    DAAL_CHECK_STATUS_VAR(st);
+    range.local(local_range, &st);
+    DAAL_CHECK_STATUS_VAR(st);
+
+    context.run(range, kernel, args, &st);
+    return st;
+}
+
+template <typename algorithmFPType>
+Status DBSCANBatchKernelUCAPI<algorithmFPType>::countPointNeighbors(const UniversalBuffer & assignments, const UniversalBuffer & pointDistances,
+                                                                    uint32_t rowId, int chunkOffset, uint32_t nRows, uint32_t numberOfChunks,
+                                                                    algorithmFPType epsP, const UniversalBuffer & queue,
+                                                                    UniversalBuffer & countersTotal, UniversalBuffer & countersNewNeighbors)
+{
+    services::Status st;
+    DAAL_ITTNOTIFY_SCOPED_TASK(compute.countPointNeighbors);
+    auto & context        = Environment::getInstance()->getDefaultExecutionContext();
+    auto & kernel_factory = context.getClKernelFactory();
+    DAAL_CHECK_STATUS_VAR(buildProgram(kernel_factory));
+    auto kernel = kernel_factory.getKernel("count_neighbors_by_type", &st);
+    DAAL_CHECK_STATUS_VAR(st);
+
+    uint32_t chunkSize = nRows / numberOfChunks + uint32_t(bool(nRows % numberOfChunks));
+    KernelArguments args(10);
+    args.set(0, assignments, AccessModeIds::read);
+    args.set(1, pointDistances, AccessModeIds::read);
+    args.set(2, rowId);
+    args.set(3, chunkOffset);
+    args.set(4, chunkSize);
+    args.set(5, nRows);
+    args.set(6, epsP);
+    args.set(7, queue, AccessModeIds::read);
+    args.set(8, countersTotal, AccessModeIds::write);
+    args.set(9, countersNewNeighbors, AccessModeIds::write);
+
+    KernelRange local_range(1, _maxWorkgroupSize);
+    KernelRange global_range(numberOfChunks * _minSubgroupSize / _maxWorkgroupSize + 1, _maxWorkgroupSize);
+
+    KernelNDRange range(2);
+    range.global(global_range, &st);
+    DAAL_CHECK_STATUS_VAR(st);
+    range.local(local_range, &st);
+    DAAL_CHECK_STATUS_VAR(st);
+
+    context.run(range, kernel, args, &st);
+    return st;
+}
+
+template <typename algorithmFPType>
+uint32_t DBSCANBatchKernelUCAPI<algorithmFPType>::sumCounters(const UniversalBuffer & counters, uint32_t numberOfChunks)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(compute.sumCounters);
+    auto cntPtr  = counters.template get<int>().toHost(ReadWriteMode::writeOnly).get();
+    uint32_t ret = 0;
+    for (uint32_t i = 0; i < numberOfChunks; i++) ret += cntPtr[i];
+    return ret;
+}
+
+template <typename algorithmFPType>
+Status DBSCANBatchKernelUCAPI<algorithmFPType>::buildProgram(ClKernelFactoryIface & kernel_factory)
+{
+    Status st;
+    auto fptype_name   = oneapi::internal::getKeyFPType<algorithmFPType>();
+    auto build_options = fptype_name;
+    build_options.add(" -D_NOISE_=-1 -D_UNDEFINED_=-2 ");
```

Comment:
`D_NOISE` is unused. Why you decided to pass an undefined label as a macro? For me, it is better to use `static const DBSCANBatchKernelUCAPI<algorithmFPType>::_undefined` field to pass it as a parameter to all kernels - this will be the one variable for both CPU and GPU code base.

Reply:
Was discussed offline